packages/compass-query-bar/src/components/option-editor.tsx

```diff
@@ -0,0 +1,157 @@
+import type { Ace } from 'ace-builds';
+
+import type { QueryOption as QueryOptionType } from '../constants/query-option-definition';
+
+const editorStyles = cx(
+  focusRingStyles,
+  css({
+    minWidth: spacing[7],
+    '&::before': {
+      position: 'absolute',
+      content: '""',
+      pointerEvents: 'none',
+      top: 3,
+      right: 3,
+      bottom: 3,
+      left: 3,
+      borderRadius: spacing[1],
+      boxShadow: 'inset 0 0 0 0 transparent',
+      transition: 'box-shadow .16s ease-in',
+    },
+    '&::after': {
+      boxShadow: '0 0 0 0 transparent',
+    },
+    borderRadius: '6px',
+    '&:hover': {
+      '&::after': {
+        boxShadow: `0 0 0 3px ${uiColors.gray.light2}`,
+        transitionTimingFunction: 'ease-out',
+      },
+    },
+    '&:focus-within': focusRingVisibleStyles,
+  })
+);
+
+const editorWithErrorStyles = css({
+  '&:focus-within': {
+    '&::before': {
+      boxShadow: 'inset 0 0 0 1px transparent',
+    },
+  },
+  '&::before': {
+    zIndex: 5,
+    boxShadow: `inset 0px 0px 0px 1px ${uiColors.red.base}`,
+  },
+});
+
+const editorSettings = {
+  useSoftTabs: true,
+  minLines: 1,
+  maxLines: 10,
+  fontSize: 12,
+  highlightActiveLine: false,
+  showGutter: false,
+};
+
+type OptionEditorProps = {
+  hasError: boolean;
+  id: string;
+  onChange: (value: string) => void;
+  onApply: () => void;
+  placeholder?: string;
+  queryOption: QueryOptionType;
+  refreshEditorAction: Listenable;
+  schemaFields?: string[];
+  serverVersion?: string;
+  value?: string;
+};
+
+export const OptionEditor: React.FunctionComponent<OptionEditorProps> = ({
+  hasError,
+  id,
+  onChange,
+  onApply,
+  placeholder,
+  refreshEditorAction,
+  schemaFields = [],
+  serverVersion = '3.6.0',
+  value = '',
+}) => {
+  const completer = useRef<typeof QueryAutoCompleter>(
+    new QueryAutoCompleter(serverVersion, EditorTextCompleter, schemaFields)
+  );
+
+  const editorRef = useRef<Ace.Editor | undefined>(undefined);
+
+  useEffect(() => {
+    const unsubscribeRefreshEditorAction = refreshEditorAction.listen(() => {
+      // Ace editor does not update the value of the editor when
+      // the container is not displayed (display: 'none').
+      // As a result, we currently listen for `subtab-changed` events
+      // which call this `refreshEditor` action. Then we perform a full ace
+      // update editor update to ensure the value is updated
+      // when the container is displayed, so that any changes to the query
+      // are reflected.
+      // More info https://github.com/securingsincity/react-ace/issues/204
+      editorRef.current?.renderer.updateFull();
+    });
+
+    return () => {
+      unsubscribeRefreshEditorAction();
+    };
+  }, [refreshEditorAction]);
+
+  useLayoutEffect(() => {
+    completer.current.update(schemaFields);
+  }, [schemaFields]);
+
+  const onApplyClicked = useCallback(() => onApply(), [onApply]);
+  const onApplyRef = useRef(onApplyClicked);
```

Comment:
You are on the right track here, but this inherits the same issue as before. Even though the `onApplyClicked` will be updated, you are only passing it to `useRef` as initial argument and so it will not update after that. The solution here is even simpler in a way, but totally not the most intuitive thing ever:

```suggestion
  const onApplyRef = useRef(onApplyClicked);
  onApplyRef.current = onApplyClicked;
```

Basically what you want is for `current` to always hold the freshest possible reference to the callback function. This is actually kinda similar to what the new (not yet added IIRC) hook `useEvent` will be for in React, you can read up on a very detailed explanation of what sort of issues it solves [here](https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md) (and if you check the "internal implementation" you'll see that it will be similar to this, just moved to the special hook, we probably don't really have enough cases to abstract it somewhere, at least for now)